File: src/main/java/registry/Registry.java
package registry;

import java.util.Collection;
import java.util.HashMap;
import java.util.function.BiConsumer;

public class Registry<T extends IRegistryEntry> {
    private HashMap<String, HashMap<String, T>> _dict = new HashMap<>();

    public T register(T v) throws DuplicateKeyException {
        RegistryName reg = v.getRegistryName();
        if (!_dict.containsKey(reg.getNamespace())) {
            _dict.put(reg.getNamespace(), new HashMap<>());
        }
        if(_dict.get(reg.getNamespace()).containsKey(reg.getName())) {
            throw new DuplicateKeyException(this, reg);
        }
        _dict.get(reg.getNamespace()).put(reg.getName(), v);
        return v;
    }

    public T get(RegistryName name) {
        if(_dict.containsKey(name.getNamespace()) && _dict.get(name.getNamespace()).containsKey(name.getName())) {
            return _dict.get(name.getNamespace()).get(name.getName());
        } else {
            return null;
        }
    }

    public int size() {
        int sz = 0;
        Collection<HashMap<String, T>> e = _dict.values();

        for(HashMap<String, T> i : e) {
            sz += i.size();
        }
        return sz;
    }

    public void forEach(String namespace, BiConsumer<String, T> op) {
        try {
            _dict.get(namespace).forEach(op);
        } catch (Exception e) {
            System.err.println("Unable to iterate over $namespace - nonexistent!");
        }
    }
}

File: src/main/kotlin/registry/Registry.kt
package registry

import kotlin.random.Random

class Registry<T: IRegistryEntry> {
    private var _dict: MutableMap<String, MutableMap<String, T>> = mutableMapOf()
    fun register(v: T): T {
        val reg = v.getRegistryName()
        if (!_dict.containsKey(reg.namespace)) {
            _dict[reg.namespace] = mutableMapOf()
        }
        if(_dict[reg.namespace]!!.containsKey(reg.name)) {
            throw DuplicateKeyException(this, reg)
        }
        _dict[reg.namespace]!![reg.name] = v
        return v
    }

    fun get(name: RegistryName): T? {
        return try {
            _dict[name.namespace]!![name.name]
        } catch (e: Exception) {
            null
        }
    }

    fun size(): Int {
        return _dict.values.fold(0, {s, e -> s + e.size})
    }

    fun iterateOverNamespace(namespace: String, op: (String, T)-> Unit) {
        try {
            _dict[namespace]!!.forEach { (s, t) -> op(s,t) }
        } catch (e: Exception) {
            Logger.logger.error("Unable to iterate over $namespace - nonexistent!")
        }
    }

    fun getRandom(namespace: String): T {
        var list = _dict[namespace]!!.entries
        return list.elementAt(Random.nextInt(list.size)).value
    }
}